src/main/java/org/rumbledb/runtime/flwor/clauses/ReturnClauseSparkIterator.java

```diff
@@ -63,18 +65,75 @@ public ReturnClauseSparkIterator(
     public JavaRDD<Item> getRDDAux(DynamicContext context) {
         RuntimeIterator expression = this.children.get(0);
         if (expression.isRDD()) {
-            throw new JobWithinAJobException(
-                    "A return clause expression cannot produce a big sequence of items for a big number of tuples, as this would lead to a data flow explosion.",
-                    getMetadata()
-            );
+            if (this.child.isDataFrame())
+                throw new JobWithinAJobException(
+                        "A return clause expression cannot produce a big sequence of items for a big number of tuples, as this would lead to a data flow explosion.",
+                        getMetadata()
+                );
+            this.tupleContext = new DynamicContext(this.currentDynamicContextForLocalExecution); // assign current
```

Comment:
this.currentDynamicContextForLocalExecution should be used only in local methods. This should probably be 'context' variable

Reply:
Yes, good catch and that was indeed one of the issues. All is fixed now.